ethcore/light/src/on_demand/mod.rs

```diff
@@ -22,174 +22,191 @@
+}
+
+impl<T: request::RequestAdapter> Future for OnResponses<T> {
+	type Item = T::Out;
+	type Error = Canceled;
+
+	fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+		self.receiver.poll().map(|async| async.map(T::extract_from))
+	}
 }
 
 /// On demand request service. See module docs for more details.
 /// Accumulates info about all peers' capabilities and dispatches
 /// requests to them accordingly.
+// lock in declaration order.
 pub struct OnDemand {
+	pending: RwLock<Vec<Pending>>,
 	peers: RwLock<HashMap<PeerId, Peer>>,
-	pending_requests: RwLock<HashMap<ReqId, Pending>>,
+	in_transit: RwLock<HashMap<ReqId, Pending>>,
 	cache: Arc<Mutex<Cache>>,
-	orphaned_requests: RwLock<Vec<Pending>>,
-	start_nonce: U256,
+	no_immediate_dispatch: bool,
 }
 
-const RECEIVER_IN_SCOPE: &'static str = "Receiver is still in scope, so it's not dropped; qed";
-
 impl OnDemand {
 	/// Create a new `OnDemand` service with the given cache.
-	pub fn new(cache: Arc<Mutex<Cache>>, account_start_nonce: U256) -> Self {
+	pub fn new(cache: Arc<Mutex<Cache>>) -> Self {
 		OnDemand {
+			pending: RwLock::new(Vec::new()),
 			peers: RwLock::new(HashMap::new()),
-			pending_requests: RwLock::new(HashMap::new()),
+			in_transit: RwLock::new(HashMap::new()),
 			cache: cache,
-			orphaned_requests: RwLock::new(Vec::new()),
-			start_nonce: account_start_nonce,
+			no_immediate_dispatch: true,
 		}
 	}
 
+	// make a test version: this doesn't dispatch pending requests
+	// until you trigger it manually.
+	#[cfg(test)]
+	fn new_test(cache: Arc<Mutex<Cache>>) -> Self {
+		let mut me = OnDemand::new(cache);
+		me.no_immediate_dispatch = true;
+
+		me
+	}
+
 	/// Request a header's hash by block number and CHT root hash.
 	/// Returns the hash.
-	pub fn hash_by_number(&self, ctx: &BasicContext, req: request::HeaderProof) -> Receiver<H256> {
-		let (sender, receiver) = oneshot::channel();
+	pub fn hash_by_number(&self, ctx: &BasicContext, req: request::HeaderProof) -> BoxFuture<H256, Canceled> {
 		let cached = {
 			let mut cache = self.cache.lock();
 			cache.block_hash(&req.num())
 		};
 
 		match cached {
-			Some(hash) => sender.send(hash).expect(RECEIVER_IN_SCOPE),
-			None => self.dispatch(ctx, Pending::HeaderProof(req, ChtProofSender::Hash(sender))),
+			Some(hash) => future::ok(hash).boxed(),
+			None => {
+				self.request(ctx, req)
+					.expect("request given fully fleshed out; qed")
+					.map(|(h, _)| h)
+					.boxed()
+			},
 		}
-		receiver
 	}
 
 	/// Request a canonical block's chain score.
 	/// Returns the chain score.
-	pub fn chain_score_by_number(&self, ctx: &BasicContext, req: request::HeaderProof) -> Receiver<U256> {
-		let (sender, receiver) = oneshot::channel();
+	pub fn chain_score_by_number(&self, ctx: &BasicContext, req: request::HeaderProof) -> BoxFuture<U256, Canceled> {
 		let cached = {
 			let mut cache = self.cache.lock();
 			cache.block_hash(&req.num()).and_then(|hash| cache.chain_score(&hash))
 		};
 
 		match cached {
```

Comment:
All these methods are removed and replaced with a more general `answer_from_cache` in #5573, would prefer to leave this refactor specifically to that PR.